### PR TITLE
add target transformer

### PR DESCRIPTION
**Pull request by lilbobbytablez**
My current use case for `skmdn` has a rather large disparity in ranges of the X and y values: [0, 25] vs [0, 300]. This seemed to affect the training. I've added the option to pass an sklearn transformer. This was in favour of using an sklearn target-tranformer as this was nearly impossible to get it play nicely with the `quantiles` parameter in `predict`.

I thought about using pytorch specific normalisation/standardisation, but as this is an sklearn package, I think it suits the user base better. In this case above, I also wanted to try `MinMaxScaler` (to keep all y-values positive) instead of the `StandardScaler` so this was another +1 for using sklearn. It's also easy to extend with CustomTransformers.

This might not be the best design but it's at least a starting point for discussion and iteration. Let me know what you think.

I'm also lining up a PR for a custom `Pipeline` object that has allowed me to apply the `StandardScaler` while also exposing the `pdf`, `cdf` and full `predict` functionality which makes this PR even more useful.

### CODE REVIEW

**Comment by koaning** [comment]
Not 100% sure if I follow why a transformer needs to be internal. Why not use scikit-learn pipelines? Something like;

```python
pipe = make_pipeline(StandardScaler(), MixtureDensityNetwork())
```

**Comment by joshdunnlime** [comment]
We may need to normalise both the target and features. The approved way to handle the target in sklearn is using the [TransformedTargetRegressor](https://scikit-learn.org/1.5/modules/generated/sklearn.compose.TransformedTargetRegressor.html) but doing this will shift a massive amount of complexity outside of the MDN class. It makes it much more complicated to calculate the pdf, cdf, predict(quantiles) and any further functionality. It's far easier to handle this internally. Yes, we could do that manually, but having watch your video "Scikit-Learn can do THAT?!" I thought why not just leverage the amazing stuff sklearn already has!

As mentioned, this is just for the target variable. I think the features should be handled using pipeline or pipeline-like* functionality.

**Comment by lilbobbytablez** [comment]
We may need to normalise both the target and features. The approved way to handle the target in sklearn is using the [TransformedTargetRegressor](https://scikit-learn.org/1.5/modules/generated/sklearn.compose.TransformedTargetRegressor.html) but doing this will shift a massive amount of complexity outside of the MDN class. It makes it much more complicated to calculate the pdf, cdf, predict(quantiles) and any further functionality. It's far easier to handle this internally. Yes, we could do that manually, but having watch your video "Scikit-Learn can do THAT?!" I thought why not just leverage the amazing stuff sklearn already has!

As mentioned, this is just for the target variable. I think the features should be handled using pipeline or pipeline-like* functionality.

Take a look at this "unscaled-moons" example. X in range [-1, 2] and y in [-10,000, 15,000]:

from skmdn import MixtureDensityEstimator

import numpy as np
from sklearn.datasets import make_moons
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt

n_samples = 1000
X_full, _ = make_moons(n_samples=n_samples, noise=0.1)
X = X_full[:, 0].reshape(-1, 1)  # Use only the first column as input
Y = X_full[:, 1].reshape(-1, 1)  # Predict the second column

# Add some noise to Y to make the problem more suitable for MDN
Y += 0.1 * np.random.randn(n_samples, 1)
Y = Y * 10000

mdn = MixtureDensityEstimator() #transformer=StandardScaler())
mdn.fit(X, Y)

means, quantiles = mdn.predict(X, quantiles=[0.01, 0.1, 0.9, 0.99], resolution=10000)

plt.scatter(X, Y)
plt.scatter(X, quantiles[:, 0], color='orange')
plt.scatter(X, quantiles[:, 1], color='green')
plt.scatter(X, quantiles[:, 2], color='green')
plt.scatter(X, quantiles[:, 3], color='orange')
plt.scatter(X, means, color='red')

**Comment by lilbobbytablez** [comment]
> Just to check, didn't I see this message with another username just before? Or am I imagining things?

Sorry, having issues switching between my old work and personal account. Should be fixed now and a more thorough example above.

**Comment by lilbobbytablez** [comment]
I think the two plots above show this fairly well, but scaling X and Y is very important for the performance of the model. I have added a notebook example `uneven-moons.ipynb` that goes over this in more detail. This could be cleaned up (removing failing cell examples) to be mini tutorial.

**Comment by koaning** [comment]
Lemme talk to some colleagues about it, may ping back later!

**Comment by lilbobbytablez** [comment]
Great. If there's a better way to achieve the same kind of thing I'd been keen to hear it!
If this is a goer then I'd be happy clean this up and start adding some tests, docs and examples.

**Comment by lilbobbytablez** [comment]
Also any thoughts on this PR? Cheers!
